academic_observatory_workflows/workflows/openalex_telescope.py

```diff
@@ -0,0 +1,465 @@
+
+        # TODO use transfer manifest instead when that is working
+        download_transfer = {"manifest": self.transfer_manifest_path_download, "bucket": self.download_bucket}
+        transform_transfer = {"manifest": self.transfer_manifest_path_transform, "bucket": self.transform_bucket}
+
+        total_count = 0
+        for transfer in [download_transfer, transform_transfer]:
+            success = False
+            prefixes = []
+            with open(transfer["manifest"], "r") as f:
+                for line in f:
+                    prefixes.append(line[1:-2])
+
+            if not prefixes:
+                continue
+
+            for i in range(max_retries):
+                if success:
+                    break
+                success, objects_count = aws_to_google_cloud_storage_transfer(
+                    aws_access_key_id,
+                    aws_secret_access_key,
+                    aws_bucket=OpenAlexTelescope.AWS_BUCKET,
+                    include_prefixes=prefixes,
+                    gc_project_id=gc_project_id,
+                    gc_bucket=transfer["bucket"],
+                    gc_bucket_path=f"telescopes/{self.dag_id}/{self.release_id}/",
+                    description=f"Transfer OpenAlex data from Airflow telescope to {transfer['bucket']}",
+                    # transfer_manifest=f"gs://{self.download_bucket}/{self.transfer_manifest_blob}"
+                )
+                total_count += objects_count
+
+            if not success:
+                raise AirflowException(f"Google Storage Transfer unsuccessful, status: {success}")
+
+        logging.info(f"Total number of objects transferred: {total_count}")
+
+    def download_transferred(self):
+        """Download the updated entities from the Google Cloud download bucket to a local directory using gsutil.
+        If the run processes the first release it will download all files. If it is a later release, it will check
+        the manifest file which tracks which entity files are modified. Only the modified files will be downloaded.
+
+        :return: None.
+        """
+        # Authenticate gcloud with service account
+        args = [
+            "gcloud",
+            "auth",
+            "activate-service-account",
+            f"--key-file" f"={os.environ['GOOGLE_APPLICATION_CREDENTIALS']}",
+        ]
+        proc: Popen = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        run_subprocess_cmd(proc, args)
+
+        logging.info(f"Downloading transferred files from Google Cloud bucket: {self.download_bucket}")
+        log_path = os.path.join(self.download_folder, "cp.log")
+        # Download all records from bucket
+        args = [
+            "gsutil",
+            "-m",
+            "-q",
+            "cp",
+            "-L",
+            log_path,
+            "-r",
+            f"gs://{self.download_bucket}/telescopes/{self.dag_id}/{self.release_id}/*",
+            self.download_folder,
+        ]
```

Comment:
Is the reasoning behind using gsutil rather than using the Google Cloud Python library to download the files because you can sync only the modified files with gsutil?

Reply:
The main reason is that it is faster to download files than using the Google Cloud Python library.
I remember doing some timings/runs with both when initially developing the ORCID telescope, which has a similar download step. I didn't write down the exact timings etc, but gsutil was more efficient.
This post doesn't have any timings either, but confirms the same thought https://stackoverflow.com/questions/44975080/is-there-any-advantage-to-using-gsutil-or-the-google-cloud-storage-api-in-produc

I'm not sure whether the Python library has the features of resuming a download using a log file and/or only copying files that are specified in a text file.

While trying to find some timing numbers I saw that there is also 'gcloud storage' in preview, which presumably has faster download/upload times, maybe something to look into.
https://cloud.google.com/blog/products/storage-data-transfer/new-gcloud-storage-enables-super-fast-data-transfers

Reply:
I see, ok that makes sense.

Reply:
Perhaps just put a comment (what you wrote above) there about why gsutil is being used, for future reference.